File: examples/python/agg-hand/handlers/request_draw.py
"""Handler for RequestDraw command."""

import sys
from pathlib import Path
from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "angzarr"))
sys.path.insert(0, str(Path(__file__).parent.parent))

from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.errors import CommandRejectedError
from angzarr_client.proto.examples import hand_pb2 as hand
from angzarr_client.proto.examples import types_pb2 as poker_types

from .state import HandState
from .game_rules import get_game_rules


def handle_request_draw(
    command_book: types.CommandBook,
    command_any: Any,
    state: HandState,
    seq: int,
) -> types.EventBook:
    """Handle RequestDraw command."""
    if not state.exists():
        raise CommandRejectedError("Hand not dealt")

    if state.status == "complete":
        raise CommandRejectedError("Hand is complete")

    cmd = hand.RequestDraw()
    command_any.Unpack(cmd)

    if not cmd.player_root:
        raise CommandRejectedError("player_root is required")

    # Check game variant supports draw
    rules = get_game_rules(state.game_variant)
    if rules.variant != poker_types.FIVE_CARD_DRAW:
        raise CommandRejectedError("Draw not supported for this game variant")

    if state.current_phase != poker_types.DRAW:
        raise CommandRejectedError("Not in draw phase")

    # Find the player
    player = None
    for p in state.players.values():
        if p.player_root == cmd.player_root:
            player = p
            break

    if not player:
        raise CommandRejectedError("Player not in hand")

    if player.has_folded:
        raise CommandRejectedError("Player has folded")

    # Validate card indices
    if len(cmd.card_indices) > 5:
        raise CommandRejectedError("Cannot discard more than 5 cards")

    for idx in cmd.card_indices:
        if idx < 0 or idx >= len(player.hole_cards):
            raise CommandRejectedError(f"Invalid card index: {idx}")

    # Check deck has enough cards
    if len(state.remaining_deck) < len(cmd.card_indices):
        raise CommandRejectedError("Not enough cards in deck")

    # Calculate new cards
    cards_discarded = len(cmd.card_indices)
    keep_indices = [
        i for i in range(len(player.hole_cards)) if i not in cmd.card_indices
    ]
    kept_cards = [player.hole_cards[i] for i in keep_indices]
    new_cards = state.remaining_deck[:cards_discarded]
    final_hand = kept_cards + new_cards

    event = hand.DrawCompleted(
        player_root=cmd.player_root,
        cards_discarded=cards_discarded,
        cards_drawn=cards_discarded,
        drawn_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    for suit, rank in new_cards:
        event.new_cards.append(poker_types.Card(suit=suit, rank=rank))

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.poker/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(
                    seconds=int(datetime.now(timezone.utc).timestamp())
                ),
            )
        ],
    )

File: examples/rust/agg-hand/src/handlers/request_draw.rs
//! RequestDraw command handler (Five Card Draw specific).

use angzarr_client::proto::examples::{BettingPhase, DrawCompleted, GameVariant, RequestDraw};
use angzarr_client::proto::{event_page, CommandBook, EventBook, EventPage};
use angzarr_client::{pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::HandState;

pub fn handle_request_draw(
    command_book: &CommandBook,
    command_any: &Any,
    state: &HandState,
    seq: u32,
) -> CommandResult<EventBook> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Hand does not exist"));
    }
    if state.is_complete() {
        return Err(CommandRejectedError::new("Hand already complete"));
    }

    // Only valid for Five Card Draw
    if state.game_variant != GameVariant::FiveCardDraw {
        return Err(CommandRejectedError::new(
            "Draw not supported in this game variant",
        ));
    }

    // Must be in draw phase
    if state.current_phase != BettingPhase::Draw {
        return Err(CommandRejectedError::new("Not in draw phase"));
    }

    let cmd: RequestDraw = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    let player = state
        .get_player(&cmd.player_root)
        .ok_or_else(|| CommandRejectedError::new("Player not in hand"))?;

    if player.has_folded {
        return Err(CommandRejectedError::new("Player has folded"));
    }

    // Validate card indices (0-4, no duplicates)
    let mut indices: Vec<i32> = cmd.card_indices.clone();
    indices.sort();
    indices.dedup();

    if indices.len() != cmd.card_indices.len() {
        return Err(CommandRejectedError::new("Duplicate card indices"));
    }

    for &idx in &indices {
        if idx < 0 || idx >= 5 {
            return Err(CommandRejectedError::new("Card index out of range (0-4)"));
        }
    }

    let cards_to_draw = indices.len();
    if cards_to_draw > state.remaining_deck.len() {
        return Err(CommandRejectedError::new("Not enough cards in deck"));
    }

    // Draw new cards from deck
    let cards_drawn = state.remaining_deck[..cards_to_draw].to_vec();

    // Build new hand: keep non-discarded cards, add new cards
    let mut new_cards = player.hole_cards.clone();
    for (i, &idx) in indices.iter().enumerate() {
        new_cards[idx as usize] = cards_drawn[i].clone();
    }

    let event = DrawCompleted {
        player_root: cmd.player_root,
        cards_discarded: cards_to_draw as i32,
        cards_drawn: cards_to_draw as i32,
        new_cards,
        drawn_at: Some(angzarr_client::now()),
    };

    let event_any = pack_event(&event, "examples.DrawCompleted");

    Ok(EventBook {
        cover: command_book.cover.clone(),
        pages: vec![EventPage {
            sequence: Some(event_page::Sequence::Num(seq)),
            event: Some(event_any),
            created_at: Some(angzarr_client::now()),
        }],
        snapshot: None,
        next_sequence: 0,
    })
}
